build.gradle

```diff
@@ -61,17 +61,33 @@ configure(subprojects.findAll { it.projectDir.toPath().startsWith(rootPath) }) {
   }
 }
 
-int prevMajor = Integer.parseInt(VersionProperties.elasticsearch.split('\\.')[0]) - 1
-String prevSnapshot = VersionProperties.elasticsearch.contains('alpha') ? '-SNAPSHOT' : ''
+// introspect all versions of ES that may be tested agains for backwards compatibility
+String currentVersion = VersionProperties.elasticsearch.minus('-SNAPSHOT')
+int prevMajor = Integer.parseInt(currentVersion.split('\\.')[0]) - 1
 File versionFile = file('core/src/main/java/org/elasticsearch/Version.java')
 List<String> versionLines = versionFile.readLines('UTF-8')
-int prevMinor = 0
+List<String> versions = []
+int prevMinorIndex = -1
+int lastPrevMinor = -1
 for (String line : versionLines) {
-  Matcher match = line =~ /\W+public static final Version V_${prevMajor}_(\d+)_.*/
+  Matcher match = line =~ /\W+public static final Version V_(\d+)_(\d+)_(\d+)(_UNRELEASED)? .*/
   if (match.matches()) {
-    prevMinor = Math.max(Integer.parseInt(match.group(1)), prevMinor)
+    int major = Integer.parseInt(match.group(1))
+    int minor = Integer.parseInt(match.group(2))
+    int bugfix = Integer.parseInt(match.group(3))
+    String versionStr = "${major}.${minor}.${bugfix}"
+    if (currentVersion != versionStr) {
+      versions.add(versionStr)
+    }
+    if (major == prevMajor && minor > lastPrevMinor) {
+      prevMinorIndex = versions.size() - 1
+      lastPrevMinor = minor
+    }
   }
 }
+versions[-1] += '-SNAPSHOT'
```

Comment:
I believe this line only makes sense if the list is sorted. We should either assert that it is sorted or walk it looking for a matching value.